File: sandbox/policy.mjs
/**
 * Phase 5 heuristic policy on BehaviorLog — evaluation configuration (b).
 */

import { decideAction } from "../action/policy.mjs";

/**
 * Map sandbox indicators plus optional classifier score to 0–100 risk.
 * @param {object} log BehaviorLog
 * @param {number} classifierRisk
 * @returns {number}
 */
export function riskFromBehaviorLog(log, classifierRisk = 0) {
  let score = Number(classifierRisk) || 0;
  const canaryHits = log?.canary_hits?.length ?? 0;
  const netEvents = log?.net?.length ?? 0;
  const fileReads = (log?.files ?? []).filter((f) => f.operation === "read").length;

  if (canaryHits > 0) {
    score = Math.max(score, 95);
  }
  if (netEvents > 0) {
    score = Math.max(score, 75);
  }
  if (fileReads > 2 && score < 50) {
    score = Math.max(score, 45);
  }
  if (log?.timeout) {
    score = Math.max(score, 55);
  }
  return Math.min(100, Math.round(score));
}

/**
 * @param {object} log BehaviorLog
 * @param {{ classifierRisk?: number, thresholds?: { quarantine: number, block: number } }} opts
 */
export function decideFromBehaviorLog(log, opts = {}) {
  const thresholds = {
    quarantine: Number(opts.thresholds?.quarantine ?? 40),
    block: Number(opts.thresholds?.block ?? 80),
  };
  const classifierRisk = Number(opts.classifierRisk ?? 0);
  const risk_score = riskFromBehaviorLog(log, classifierRisk);
  const action = decideAction(risk_score, thresholds);

  const reasons = [];
  if ((log?.canary_hits?.length ?? 0) > 0) {
    reasons.push(
      `canary_hits=${log.canary_hits.map((h) => h.canary_id).join(",")}`,
    );
  }
  if ((log?.net?.length ?? 0) > 0) {
    reasons.push(`net_attempts=${log.net.length}`);
  }
  if (log?.timeout) {
    reasons.push("sandbox_timeout");
  }
  if (classifierRisk > 0) {
    reasons.push(`classifier_risk=${classifierRisk}`);
  }

  return {
    config: "b",
    pipeline: "classifier+sandbox",
    risk_score,
    action,
    behavior_log_version: log?.behavior_log_version ?? "1.0.0",
    timeout: Boolean(log?.timeout),
    canary_hits: log?.canary_hits ?? [],
    justification:
      reasons.length > 0
        ? `Sandbox indicators: ${reasons.join("; ")} → risk_score=${risk_score}.`
        : `No sandbox tripwires; risk_score=${risk_score}.`,
  };
}

File: eval/policy.py
"""Policy helpers for evaluation configs (a)(b)(c)."""

from __future__ import annotations

from typing import Any

DEFAULT_THRESHOLDS = {"quarantine": 40, "block": 80}


def risk_from_features(features: dict[str, Any]) -> int:
    raw = float(features.get("suspicion_score") or 0)
    score = min(100, round(raw * 10))
    if features.get("unparseable") and raw < 3:
        score = min(score, 25)
    return int(score)


def decide_action(risk_score: int, thresholds: dict[str, int]) -> str:
    if risk_score > thresholds["block"]:
        return "block"
    if risk_score >= thresholds["quarantine"]:
        return "quarantine"
    return "allow"


def risk_from_behavior_log(
    log: dict[str, Any] | None, classifier_risk: int = 0
) -> int:
    score = int(classifier_risk or 0)
    log = log or {}
    canary_hits = len(log.get("canary_hits") or [])
    net_events = len(log.get("net") or [])
    file_reads = len(
        [f for f in (log.get("files") or []) if f.get("operation") == "read"]
    )

    if canary_hits > 0:
        score = max(score, 95)
    if net_events > 0:
        score = max(score, 75)
    if file_reads > 2 and score < 50:
        score = max(score, 45)
    if log.get("timeout"):
        score = max(score, 55)
    return min(100, int(score))


def risk_from_triage(malicious_probability: float, label: str) -> int:
    if label == "benign":
        return int(min(35, malicious_probability * 30))
    if label == "suspicious":
        return int(40 + malicious_probability * 40)
    return int(70 + malicious_probability * 30)


def action_is_positive(action: str) -> bool:
    return action in {"quarantine", "block"}
